README.md
Language: [English](README.md) | [中文简体](README-ZH.md)
# switch_when

A function library that provides a more advanced switch, does not restrict Case expressions to be constants, similar to kotlin's when method.

## Getting Started

### Add dependency

```yaml
dependencies:
  switch_when: ^0.0.1 #please use the latest version on pub
```
 
### Use Library
1. import package
 Where you need to use the Library library, import the package
```dart
import 'package:switch_when/switch_when.dart';
```

2. Use functions
Different methods are used according to different situations. For general situations, you can use the when method or whenValue method. If you want the type of the returned result to be null safe, you can use the whenSafe method or whenValueSafe method.
```dart
  @override
  void initState() {
    _curtainAnimationImage = TweenImageWidget(
      ImagesEntry(1, 20, "equipmentcontrol_pic_curtain%s".toAssetImg()),
      durationMilliseconds: 5000,
      repeat: false,
    );
    super.initState();
  }
```


**Functions:**
```dart
//Not Safe
T? whenString<T>(String value, Map<String, ValueGetter<T>> conditionMap)
T? whenInt<T>(int value, Map<int, ValueGetter<T>> conditionMap)
T? whenDouble<T>(double value, Map<double, ValueGetter<T>> conditionMap) 
T? whenBool<T>(bool value, List<Tuple2<bool, ValueGetter<T>>> conditionList)
T? whenValue<V, T>(V value, Map<V, ValueGetter<T>> conditionMap)
T? when<T>(Map<bool, ValueGetter<T>> conditionMap)
T? whenTrue<T>(Map<ValueGetter<bool>, ValueGetter<T>> conditionMap)

//Safe
T whenStringSafe<T>(String value, Map<String, ValueGetter<T>> conditionMap, {ValueGetter<T>? defaultValue}) 
T whenIntSafe<T>(int value, Map<int, ValueGetter<T>> conditionMap, {ValueGetter<T>? defaultValue})
T whenDoubleSafe<T>(double value, Map<double, ValueGetter<T>> conditionMap, {ValueGetter<T>? defaultValue}) 
T? whenBoolSafe<T>(bool value, List<Tuple2<bool, ValueGetter<T>>> conditionList, {ValueGetter<T>? defaultValue}) 
T whenValueSafe<V, T>(V value, Map<V, ValueGetter<T>> conditionMap, {ValueGetter<T>? defaultValue}) 
T whenSafe<T>(Map<bool, ValueGetter<T>> conditionMap, {ValueGetter<T>? defaultValue})
T whenTrueSafe<T>(Map<ValueGetter<bool>, ValueGetter<T>> conditionMap, {ValueGetter<T>? defaultValue}) 
```


<details>
  <summary>T? whenString<T>(String value, Map<String, ValueGetter<T>> conditionMap)</summary>
<p>Used to replace the [switch] method, because in some scenarios, an error warning of Case expressions must be constant;<br>
If there is [value] in [Map.keys] of [conditionMap], execute its corresponding [ValueGetter] method
</p>

example:
```dart
 int? index = whenString<int>("banana🍌", {
   "water" + "melon": () {
     return 1;
   },
   "apple": () {
     return 2;
   },
   "orange": () {
     return 3 ;
   },
   "banana" + "🍌": () {
     return 4;
   },
   "grape": () {
     return 5;
   },
 });
```
</details>


<details>
  <summary>T? whenInt<T>(int value, Map<int, ValueGetter<T>> conditionMap)</summary>
<p>Used to replace the [switch] method, because in some scenarios, an error warning of Case expressions must be constant;<br>
If there is [value] in [Map.keys] of [conditionMap], execute its corresponding [ValueGetter] method
</p>

example:
```dart
 String? status = whenInt<String>(1, {
   1: () {
     return "good";
   },
   1 + 1: () {
     return "nice";
   },
   int.parse("3"): () {
     return "better";
   },
 });
```
</details>

<details>
  <summary>T? whenDouble<T>(double value, Map<double, ValueGetter<T>> conditionMap)</summary>
<p>Used to replace the [switch] method, because in some scenarios, an error warning of Case expressions must be constant;<br>
If there is [value] in [Map.keys] of [conditionMap], execute its corresponding [ValueGetter] method
</p>

example:
```dart
 String? status = whenDouble<String>(2.0, {
   0.1: () {
     return "good";
   },
   1.0 + 1: () {
     return "nice";
   },
   double.parse("3.2"): () {
     return "better";
   },
 });
```
</details>


<details>
  <summary>T? whenBool<T>(bool value, List<Tuple2<bool, ValueGetter<T>>> conditionList)</summary>
<p> Used to replace the switch method, because some scenes use [switch] to cause an error warning of Case expressions must be constant.<br>
If there is [value] in [Tuple2.item1] of [conditionList], execute its corresponding [ValueGetter] method
</p>

example:
```dart
  double? degree = whenBool<double>(false, [
    Tuple2(
      "is Long String".length > 10,
      () {
        return 0.0;
      },
    ),
    Tuple2(
      100 / 10 == 0,
      () {
        return 1.0;
      },
    ),
    Tuple2(
      "apple".contains("a"),
      () {
        return 2.0;
      },
    ),
  ]);
  return degree;
```
</details>


<details>
  <summary>T? whenValue<V, T>(V value, Map<V, ValueGetter<T>> conditionMap)</summary>
<p>The super evolution version💖💖💖 of [switch] method, all basic types of values can be compared, including [List], [Map], [Set], and [Iterable].<br>
All need do is [value] in [Map.keys] of [conditionMap], its corresponding [ValueGetter] method will be executed
</p>

example:
```dart
 String? kind = whenValue<List, String>(
   ["apple", "orange"],
   {
     ["cat", "dog"]: () {
       return "pets";
     },
     ["apple", "orange"]: () {
       return "fruits";
     },
     ["red", "white", "black"]: () {
       return "colors";
     },
   },
 );
```
</details>


<details>
  <summary>T? when<T>(Map<bool, ValueGetter<T>> conditionMap)</summary>
<p>The when function of the Kotlin version of the method [switch].<br>
 As long as the first true is found in [Map.keys] of [conditionMap], its corresponding [ValueGetter] method will be executed immediately and the relative value will be returned.<br>
 If it is not found, it will return null; if you need a default value, you can add a MapEntry with a key equal to true at the end of the Map
</p>

example:
```dart
 String? winner = when<String>({
   "Dart is Language".contains("UI"): () {
     return "Flutter";
   },
   "Flutter is UI Framework".contains("UI"): () {
     return "Flutter";
   },
 });
```
</details>


<details>
  <summary>T? whenTrue<T>(Map<ValueGetter<bool>, ValueGetter<T>> conditionMap)</summary>
<p>The when function of the Kotlin version of the method [switch]，its conditional expression will be calculated。.<br>
 As long as [conditionMap] to [Map.keys] appears first execution result is true, it will immediately perform corresponding [ValueGetter] method, and an opposite return value.<br>
 If it is not found, it will return null; if you need a default value, you can add a MapEntry with a key equal to true at the end of the Map
</p>

example:
```dart
String? something = whenTrue<String>({
      () {
    if (1 + 100 * 1000 < 2000) {
      return false;
    } else if ("Who is my lovely baby?".length > 10) {
      return true;
    } else {
      return false;
    }
  }: () {
    return "Test OK";
  },
      () {
    return int.tryParse("3.14*") != null;
  }: () {
    return "PI get";
  }
});
```
</details>

### Example Demo:

```dart
  testWhenValue() {
    String? kind = whenValue<List, String>(
      ["apple", "orange"],
      {
        ["cat", "dog"]: () {
          return "pets";
        },
        ["apple", "orange"]: () {
          return "fruits";
        },
        ["red", "white", "black"]: () {
          return "colors";
        },
      },
    );
    return kind;
  }

  testWhenString() {
    int howManyFruits = 2;
    int? index = whenString<int>("banana🍌", {
      "water" + "melon".more(): () {
        return 1;
      },
      "apple".more(): () {
        return 2;
      },
      howManyFruits <= 1 ? "orange" : "oranges": () {
        return 3;
      },
      "banana" + "🍌": () {
        return 4;
      },
      "grape": () {
        return 5;
      },
    });
    return index;
  }

  testWhen() {
    String? winner = when<String>({
      "Dart is Language".contains("UI"): () {
        return "Flutter";
      },
      "Flutter is UI Framework".contains("UI"): () {
        return "Flutter";
      },
    });
    return winner;
  }

  testWhenDoubleSafe() {
    String status = whenDoubleSafe<String>(
      2.1,
      {
        0.1: () {
          return "good";
        },
        1.0 + 1: () {
          return "nice";
        },
        double.parse("3.2"): () {
          return "better";
        },
      },
      defaultValue: () => "other",
    );
    expect(status, "other");
  };
```

#### Demo Pictrue
![demo](demo.png)






CHANGELOG.md
## [0.0.2] - Release date: 2021-07-05.

* Second version: Add whenBool & whenBoolSafe Method; The library path is changed from index to switch_when.

## [0.0.1] - Release date: 2021-07-02.

* First version: initial release.


test/switch_when_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:switch_when/switch_when.dart';
import 'package:tuple/tuple.dart';

extension Stringx on String {
  String more() {
    return "$this more";
  }
}

void main() {
  group('when_extensions', () {
    test('testWhenString', () {
      int howManyFruits = 2;
      int? index = whenString<int>("banana🍌", {
        "water" + "melon".more(): () {
          return 1;
        },
        "apple".more(): () {
          return 2;
        },
        howManyFruits <= 1 ? "orange" : "oranges": () {
          return 3;
        },
        "banana" + "🍌": () {
          return 4;
        },
        "grape": () {
          return 5;
        },
      });

      expect(index, 4);
    });

    test('testWhenInt', () {
      String? status = whenInt<String>(3, {
        1: () {
          return "good";
        },
        1 + 1: () {
          return "nice";
        },
        int.parse("3"): () {
          return "better";
        },
      });
      expect(status, "better");
    });

    test('testWhenDouble', () {
      String? status = whenDouble<String>(2.0, {
        0.1: () {
          return "good";
        },
        1.0 + 1: () {
          return "nice";
        },
        double.parse("3.2"): () {
          return "better";
        },
      });
      expect(status, "nice");
    });

    test('testWhenDouble', () {
      double? degree = whenBool<double>(false, [
        Tuple2(
          "is Long String".length > 10,
          () {
            return 0.0;
          },
        ),
        Tuple2(
          100 / 10 == 0,
          () {
            return 1.0;
          },
        ),
        Tuple2(
          "apple".contains("a"),
          () {
            return 2.0;
          },
        ),
      ]);
      expect(degree, 1.0);
    });

    test('testWhenValue', () {
      String? kind = whenValue<List, String>(
        ["apple", "orange"],
        {
          ["cat", "dog"]: () {
            return "pets";
          },
          ["apple", "orange"]: () {
            return "fruits";
          },
          ["red", "white", "black"]: () {
            return "colors";
          },
        },
      );
      expect(kind, "fruits");
    });

    test('testWhen', () {
      String? winner = when<String>({
        "Dart is Language".contains("UI"): () {
          return "Flutter";
        },
        "Flutter is UI Framework".contains("UI"): () {
          return "Flutter";
        },
      });
      expect(winner, "Flutter");
    });

    test('testWhenTrue', () {
      String? something = whenTrue<String>({
        () {
          if (1 + 100 * 1000 < 2000) {
            return false;
          } else if ("Who is my lovely baby?".length > 100) {
            return true;
          } else {
            return false;
          }
        }: () {
          return "Test OK";
        },
        () {
          return int.tryParse("3.14*") != null;
        }: () {
          return "PI get";
        }
      });
      expect(something, null);
    });
  });

  group('when_safe_extensions', () {
    test('testWhenStringSafe', () {
      int howManyFruits = 2;
      int index = whenStringSafe<int>(
        "🍌🍌🍌",
        {
          "water" + "melon".more(): () {
            return 1;
          },
          "apple".more(): () {
            return 2;
          },
          howManyFruits <= 1 ? "orange" : "oranges": () {
            return 3;
          },
          "banana" + "🍌": () {
            return 4;
          },
          "grape": () {
            return 5;
          },
        },
        defaultValue: () => 0,
      );

      expect(index, 0);
    });

    test('whenIntSafe', () {
      String status = whenIntSafe<String>(3, {
        1: () {
          return "good";
        },
        1 + 1: () {
          return "nice";
        },
        int.parse("3"): () {
          return "better";
        },
      });
      expect(status, "better");
    });

    test('whenDoubleSafe', () {
      String status = whenDoubleSafe<String>(
        2.1,
        {
          0.1: () {
            return "good";
          },
          1.0 + 1: () {
            return "nice";
          },
          double.parse("3.2"): () {
            return "better";
          },
        },
        defaultValue: () => "other",
      );
      expect(status, "other");
    });

    test('testWhenValue', () {
      String kind = whenValueSafe<List, String>(
        ["apple", "orange"],
        {
          ["cat", "dog"]: () {
            return "pets";
          },
          ["apple", "orange"]: () {
            return "fruits";
          },
          ["red", "white", "black"]: () {
            return "colors";
          },
        },
      );
      expect(kind, "fruits");
    });

    test('testWhen', () {
      String winner = whenSafe<String>({
        "Dart is Language".contains("UI"): () {
          return "Flutter";
        },
        "Flutter is UI Framework".contains("UI"): () {
          return "Flutter";
        },
      });
      expect(winner, "Flutter");
    });

    test('testWhenTrue', () {
      String something = whenTrueSafe<String>(
        {
          () {
            if (1 + 100 * 1000 < 2000) {
              return false;
            } else if ("Who is my lovely baby?".length > 100) {
              return true;
            } else {
              return false;
            }
          }: () {
            return "Test OK";
          },
          () {
            return int.tryParse("3.14*") != null;
          }: () {
            return "PI get";
          }
        },
        defaultValue: () => "defaultValue",
      );
      expect(something, "defaultValue");
    });
  });
}


lib/src/when_extensions.dart
import 'package:collection/collection.dart';
import 'package:flutter/foundation.dart';
import 'package:tuple/tuple.dart';
import 'utils.dart';

///English: Used to replace the [switch] method, because in some scenarios, an error warning of Case expressions must be constant;<br>
///If there is [value] in [Map.keys] of [conditionMap], execute its corresponding [ValueGetter] method
///
///中文: 用于取代switch方法，因为有些场景使用[switch]会出现Case expressions must be constant.的错误警告；<br>
///如果[conditionMap]的[Map.keys]中有[value]的话，执行其对应的[ValueGetter]方法
///
///
///example:
///```dart
/// int? index = whenString<int>("banana🍌", {
///   "water" + "melon": () {
///     return 1;
///   },
///   "apple": () {
///     return 2;
///   },
///   "orange": () {
///     return 3 ;
///   },
///   "banana" + "🍌": () {
///     return 4;
///   },
///   "grape": () {
///     return 5;
///   },
/// });
/// ```
T? whenString<T>(String value, Map<String, ValueGetter<T>> conditionMap) {
  return conditionMap[value]?.call();
}

///English: Used to replace the [switch] method, because in some scenarios, an error warning of Case expressions must be constant;<br>
///If there is [value] in [Map.keys] of [conditionMap], execute its corresponding [ValueGetter] method
///
///中文: 用于取代switch方法，因为有些场景使用[switch]会出现Case expressions must be constant.的错误警告；<br>
///如果[conditionMap]的[Map.keys]中有[value]的话，执行其对应的[ValueGetter]方法
///
///
///example:
///```dart
/// String? status = whenInt<String>(1, {
///   1: () {
///     return "good";
///   },
///   1 + 1: () {
///     return "nice";
///   },
///   int.parse("3"): () {
///     return "better";
///   },
/// });
/// ```
T? whenInt<T>(int value, Map<int, ValueGetter<T>> conditionMap) {
  return conditionMap[value]?.call();
}

///English: Used to replace the [switch] method, because in some scenarios, an error warning of Case expressions must be constant;<br>
///If there is [value] in [Map.keys] of [conditionMap], execute its corresponding [ValueGetter] method
///
///中文: 用于取代switch方法，因为有些场景使用[switch]会出现Case expressions must be constant.的错误警告；<br>
///如果[conditionMap]的[Map.keys]中有[value]的话，执行其对应的[ValueGetter]方法
///
///
///example:
///```dart
/// String? status = whenDouble<String>(2.0, {
///   0.1: () {
///     return "good";
///   },
///   1.0 + 1: () {
///     return "nice";
///   },
///   double.parse("3.2"): () {
///     return "better";
///   },
/// });
/// ```
T? whenDouble<T>(double value, Map<double, ValueGetter<T>> conditionMap) {
  return conditionMap[value]?.call();
}

///English: Used to replace the switch method, because some scenes use [switch] to cause an error warning of Case expressions must be constant.<br>
/// If there is [value] in [Tuple2.item1] of [conditionList], execute its corresponding [ValueGetter] method
///
///中文: 用于取代switch方法，因为有些场景使用[switch]会出现Case expressions must be constant.的错误警告；<br>
/// 如果[conditionList]的[Tuple2.item1]中有[value]的话，执行其对应的[ValueGetter]方法
///
///
///example:
///```dart
/// double? degree = whenBool<double>(false, [
///   Tuple2(
///     "is Long String".length > 10,
///         () {
///       return 0.0;
///     },
///   ),
///   Tuple2(
///     100 / 10 == 0,
///         () {
///       return 1.0;
///     },
///   ),
///   Tuple2(
///     "apple".contains("a"),
///         () {
///       return 2.0;
///     },
///   ),
/// ]);
/// return degree;
/// ```
T? whenBool<T>(bool value, List<Tuple2<bool, ValueGetter<T>>> conditionList) {
  for (var conditionTuple in conditionList) {
    if (conditionTuple.item1 == value) {
      return conditionTuple.item2();
    }
  }
  return null;
}

/// English: The super evolution version💖💖💖 of [switch] method, all basic types of values can be compared, including [List], [Map], [Set], and [Iterable].<br>
/// All need do is [value] in [Map.keys] of [conditionMap], its corresponding [ValueGetter] method will be executed
///
/// 中文: 方法[switch]的超级进化版本💖💖💖，所有基本类型的value都可以比较，包括[List],[Map]，[Set]，以及[Iterable]。<br>
/// 只要[conditionMap]的[Map.keys]中有[value]的话，就会执行其对应的[ValueGetter]方法
///
///
///example:
///```dart
/// String? kind = whenValue<List, String>(
///   ["apple", "orange"],
///   {
///     ["cat", "dog"]: () {
///       return "pets";
///     },
///     ["apple", "orange"]: () {
///       return "fruits";
///     },
///     ["red", "white", "black"]: () {
///       return "colors";
///     },
///   },
/// );
/// ```
T? whenValue<V, T>(V value, Map<V, ValueGetter<T>> conditionMap) {
  return conditionMap.entries
      .firstOrNullWhere((element) =>
          const DeepCollectionEquality().equals(element.key, value))
      ?.value();
}

/// English: The when function of the Kotlin version of the method [switch].<br>
/// As long as the first true is found in [Map.keys] of [conditionMap], its corresponding [ValueGetter] method will be executed immediately and the relative value will be returned.<br>
/// If it is not found, it will return null; if you need a default value, you can add a MapEntry with a key equal to true at the end of the Map
///
/// Special attention!!! Due to the characteristics of [Map], you need to ensure that [Map.keys] in [conditionMap] does not have multiple true values.
/// If you cannot guarantee it, it is recommended to use the [whenTrue] or [whenBool] method to avoid program problems
///
/// 中文: 方法[switch]的kotlin版本的when函数.<br>
/// 只要在[conditionMap]的[Map.keys]中发现第一个true,就会立刻执行其对应的[ValueGetter]方法，并返回相对的值.<br>
/// 如果没有找到的话，会返回null;如果需要默认值，可以在Map中最后加入一个key等于true的MapEntry
///
/// 特别注意!!! 由于[Map]的特性，需要确保[conditionMap]中的[Map.keys]不会出现多个true的值。
/// 如果不能保证，建议使用[whenTrue]或者[whenBool]方法，这样可以避免程序出现问题
///
///
///example:
///```dart
/// String? winner = when<String>({
///   "Dart is Language".contains("UI"): () {
///     return "Flutter";
///   },
///   "Flutter is UI Framework".contains("UI"): () {
///     return "Flutter";
///   },
/// });
/// ```
T? when<T>(Map<bool, ValueGetter<T>> conditionMap) {
  for (var element in conditionMap.entries) {
    if (element.key) {
      return element.value();
    }
  }
  return null;
}

/// English: The when function of the Kotlin version of the method [switch]，its conditional expression will be calculated。.<br>
/// As long as [conditionMap] to [Map.keys] appears first execution result is true, it will immediately perform corresponding [ValueGetter] method, and an opposite return value.<br>
/// If it is not found, it will return null; if you need a default value, you can add a MapEntry with a key equal to true at the end of the Map
///
/// 中文: 方法[switch]的kotlin版本的when函数.<br>，其条件表达式会进行计算。
/// 只要在[conditionMap]的[Map.keys]中第一个出现执行结果为true,就会立刻执行其对应的[ValueGetter]方法，并返回相对的值.<br>
/// 如果没有找到的话，会返回null;如果需要默认值，可以在Map中最后加入一个key等于true的MapEntry
///
///
///example:
///```dart
///String? something = whenTrue<String>({
///      () {
///    if (1 + 100 * 1000 < 2000) {
///      return false;
///    } else if ("Who is my lovely baby?".length > 10) {
///      return true;
///    } else {
///      return false;
///    }
///  }: () {
///    return "Test OK";
///  },
///      () {
///    return int.tryParse("3.14*") != null;
///  }: () {
///    return "PI get";
///  }
///});
/// ```
T? whenTrue<T>(Map<ValueGetter<bool>, ValueGetter<T>> conditionMap) {
  for (var element in conditionMap.entries) {
    if (element.key.call()) {
      return element.value();
    }
  }
  return null;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:switch_when/switch_when.dart';
import 'package:code_editor/code_editor.dart';
import 'package:tuple/tuple.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'switch_when Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Switch When Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

extension Stringx on String {
  String more() {
    return "$this more";
  }
}

class _MyHomePageState extends State<MyHomePage> {
  testWhenString() {
    int howManyFruits = 2;
    int? index = whenString<int>("banana🍌", {
      "water" + "melon".more(): () {
        return 1;
      },
      "apple".more(): () {
        return 2;
      },
      howManyFruits <= 1 ? "orange" : "oranges": () {
        return 3;
      },
      "banana" + "🍌": () {
        return 4;
      },
      "grape": () {
        return 5;
      },
    });
    return index;
  }

  testWhenInt() {
    String? status = whenInt<String>(1, {
      1: () {
        return "good";
      },
      1 + 1: () {
        return "nice";
      },
      int.parse("3"): () {
        return "better";
      },
    });
    return status;
  }

  testWhenDouble() {
    String? status = whenDouble<String>(2.0, {
      0.1: () {
        return "good";
      },
      1.0 + 1: () {
        return "nice";
      },
      double.parse("3.2"): () {
        return "better";
      },
    });
    return status;
  }

  testWhenBool() {
    double? degree = whenBool<double>(false, [
      Tuple2(
        "is Long String".length > 10,
        () {
          return 0.0;
        },
      ),
      Tuple2(
        100 / 10 == 0,
        () {
          return 1.0;
        },
      ),
      Tuple2(
        "apple".contains("a"),
        () {
          return 2.0;
        },
      ),
    ]);
    return degree;
  }

  testWhenValue() {
    String? kind = whenValue<List, String>(
      ["apple", "orange"],
      {
        ["cat", "dog"]: () {
          return "pets";
        },
        ["apple", "orange"]: () {
          return "fruits";
        },
        ["red", "white", "black"]: () {
          return "colors";
        },
      },
    );
    return kind;
  }

  testWhen() {
    String? winner = when<String>({
      "Dart is Language".contains("UI"): () {
        return "Flutter";
      },
      "Flutter is UI Framework".contains("UI"): () {
        return "Flutter";
      },
    });
    return winner;
  }

  testWhenTrue() {
    String? something = whenTrue<String>({
      () {
        if (1 + 100 * 1000 < 2000) {
          return false;
        } else if ("Who is my lovely baby?".length > 10) {
          return true;
        } else {
          return false;
        }
      }: () {
        return "Test OK";
      },
      () {
        return int.tryParse("3.14*") != null;
      }: () {
        return "PI get";
      }
    });
    return something;
  }

  Map<String, Tuple2<String, Function>> get data => {
        "WhenString": Tuple2(
          """
    int howManyFruits = 2;
    int? index = whenString<int>("banana🍌", {
      "water" + "melon".more(): () {
        return 1;
      },
      "apple".more(): () {
        return 2;
      },
      howManyFruits <= 1 ? "orange" : "oranges": () {
        return 3;
      },
      "banana" + "🍌": () {
        return 4;
      },
      "grape": () {
        return 5;
      },
    });
    """,
          testWhenString,
        ),
        "whenInt": Tuple2(
          """
      String? status = whenInt<String>(1, {
      1: () {
        return "good";
      },
      1 + 1: () {
        return "nice";
      },
      int.parse("3"): () {
        return "better";
      },
    });
    """,
          testWhenInt,
        ),
        "whenDouble": Tuple2(
          """
      String? status = whenDouble<String>(2.0, {
      0.1: () {
        return "good";
      },
      1.0 + 1: () {
        return "nice";
      },
      double.parse("3.2"): () {
        return "better";
      },
    });
    """,
          testWhenDouble,
        ),
        "whenBool": Tuple2(
          """
      double? degree = whenBool<double>(false, [
      Tuple2(
        "is Long String".length > 10,
        () {
          return 0.0;
        },
      ),
      Tuple2(
        100 / 10 == 0,
        () {
          return 1.0;
        },
      ),
      Tuple2(
        "apple".contains("a"),
        () {
          return 2.0;
        },
      ),
    ]);
    return degree;
    """,
          testWhenBool,
        ),
        "whenValue": Tuple2(
          """
       String? kind = whenValue<List, String>(
      ["apple", "orange"],
      {
        ["cat", "dog"]: () {
          return "pets";
        },
        ["apple", "orange"]: () {
          return "fruits";
        },
        ["red", "white", "black"]: () {
          return "colors";
        },
      },
    );
    """,
          testWhenValue,
        ),
        "when": Tuple2(
          """
    String? winner = when<String>({
      "Dart is Language".contains("UI"): () {
        return "Flutter";
      },
      "Flutter is UI Framework".contains("UI"): () {
        return "Flutter";
      },
    });
    """,
          testWhen,
        ),
        "whenTrue": Tuple2(
          """
    String? something = whenTrue<String>({
      () {
        if (1 + 100 * 1000 < 2000) {
          return false;
        } else if ("Who is my lovely baby?".length > 10) {
          return true;
        } else {
          return false;
        }
      }: () {
        return "Test OK";
      },
      () {
        return int.tryParse("3.14*") != null;
      }: () {
        return "PI get";
      }
    });
    """,
          testWhenTrue,
        ),
      };

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: ListView(
          children: data.entries
              .map((method) => Column(
                    children: [
                      IgnorePointer(
                        child: CodeEditor(
                          model: EditorModel(
                            styleOptions: EditorModelStyleOptions(
                              heightOfContainer:
                                  method.value.item1.split("\n").length * 18,
                              fontSize: 12,
                              lineHeight: 1.35,
                              padding: EdgeInsets.symmetric(
                                  vertical: 8, horizontal: 15),
                            ),
                            files: [
                              FileEditor(
                                  name: method.key,
                                  language: "dart",
                                  code: method.value.item1),
                            ],
                          ),
                          edit: false,
                          disableNavigationbar: false,
                        ),
                      ),
                      Container(
                        padding:
                            EdgeInsets.symmetric(horizontal: 18, vertical: 10),
                        color: Colors.deepPurple,
                        alignment: Alignment.centerLeft,
                        child: Text(
                          "run result=> ${method.value.item2()}",
                          style: TextStyle(color: Colors.white),
                        ),
                      ),
                    ],
                  ))
              .toList(),
        ),
      ),
    );
  }
}
